### PR TITLE
fix(strapi-v4): Strapi-v4-data-provider-not-working-with-nested-conditional-operators

### LINKED ISSUE #5479: [BUG] Strapi v4 data provider does not working with nested filters $or and $and operator
### Describe the bug

image
image

### Steps To Reproduce

```ts
import { generateFilter } from "@refinedev/strapi-v4";

const filters = [
  {
    operator: "or",
    value: [
      {
        field: "affect_global",
        operator: "eq",
        value: true,
      },
      {
        operator: "and",
        value: [
          {
            field: "studio.id",
            operator: "eq",
            value: 2,
          },
          {
            field: "is_primary",
            operator: "eq",
            value: true,
          },
        ],
      },
    ],
  },
];

const query = generateFilter(filters);
```

### Expected behavior

`filters[$or][0][affect_global][$eq]=true&filters[$or][1][$and][0][studio][id][$eq]=2&filters[$or][1][$and][1][is_primary][$eq]=true`

### Screenshot

_No response_

### Desktop

_No response_

### Mobile

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by vidurkhanal**
fix: Looks into solving [Issue #5479](https://github.com/refinedev/refine/issues/5479)

Explanation:

Current implementation of an utility function in strapi-v4 package, generateFilter(), doesn't handle all the edge cases, as pointed out by the issue owner.  The original code doesn't handle the case where the value of a conditional filter may be another conditional filter. It assumes that value of a conditional filter will be a logical filter.

I have chosen a recursive parsing approach to solve this and do believe produce expected raw string. However, the raw string after being parsed and stringified through "qs" adds unnecessary square braces to the URL params. Attached is the screenshot of the issue. The console.log is me logging the raw string before parsing and stringifying. 

Screenshot 2024-01-15 at 10 00 43 AM

### Test plan (required)

WIP: Added a new test case when there are recursive conditional filters.

### Closing issues
closes #5479 

### Self Check before Merge

Please check all items below before review.

-   [.] Corresponding issues are created/updated or not needed
-   [.] Docs are updated/provided or not needed
-   [.] Examples are updated/provided or not needed
-   [.] TypeScript definitions are updated/provided or not needed
-   [.] Tests are updated/provided or not needed
-   [.] Changesets are provided or not needed

### CODE REVIEW

**Comment by alicanerdurmaz** [comment]
Hello @vidurkhanal, thanks for the PR. this is looks neat 🚀

Can you provide a [changeset](https://refine.dev/docs/guides-concepts/contributing/#creating-a-changeset) please?

File: packages/strapi-v4/src/utils/generateFilter.ts [hunk 1]
```diff
@@ -14,49 +14,78 @@ export const generateNestedFilterField = (field: string) => {
...
-                        field,
-                    )}[$${mapedOperator}]=${value}`;
-                });
+                // the filter is a conditional filter
+                rawQuery += generateConditionalFilter(filter);
             }
         });
     }
 
+    console.log(rawQuery);
```
Comment (alicanerdurmaz) [inline packages/strapi-v4/src/utils/generateFilter.ts:88]:
```suggestion
    
```

**Review by alicanerdurmaz — CHANGES_REQUESTED** [review]

**Comment by alicanerdurmaz** [comment]
@vidurkhanal It looks like you're still working on it, so can you pull PR to the draft, please?

**Comment by alicanerdurmaz** [comment]
> Hey @alicanerdurmaz , I was actually looking for suggestion from the team. The raw string after being parsed and stringified through "qs" adds unnecessary square braces to the URL params. So I was wondering if some one from the team could take a look. The screenshot is attached as a reference when console logging the raw string and stringified url by 'qs'

Oh, I see. Thanks for the detailed explanation. I will try to find a solution when I find time.

**Comment by rilrom** [comment]
@vidurkhanal

Adding a `depth` on `parse` resolves your test case.

`const parsedQuery = parse(rawQuery, { depth: 50 });`

I recommend determining what maximum depth is suitable for our use-case.

**Comment by vidurkhanal** [comment]
> @vidurkhanal
> 
> Adding a `depth` on `parse` resolves your test case.
> 
> `const parsedQuery = parse(rawQuery, { depth: 50 });`
> 
> I recommend determining what maximum depth is suitable for our use-case. The [qs documentation](https://github.com/ljharb/qs?tab=readme-ov-file#parsing-objects) provides some good explanations about depth.

Thanks @rilrom, I will look into that immediately.

**Comment by rilrom** [comment]
No worries @vidurkhanal.

Feel free to reach out if you'd like me to adjust the solution and put together a changeset for you.

**Comment by BatuhanW** [comment]
Closing the PR due to inactivity. @rilrom feel free to send a PR.

**Pull request closed without merging**